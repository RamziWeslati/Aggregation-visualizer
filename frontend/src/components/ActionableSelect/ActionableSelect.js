import React from "react";
import PropTypes from 'prop-types';

import { NativeSelect, Grid } from "@material-ui/core";
// import CancelIcon from "@material-ui/icons/Cancel";
import CancelPresentationTwoToneIcon from '@material-ui/icons/CancelPresentationTwoTone';

const ActionableSelect = ({ data, placeholder, action }) => {
  /**Component to render select that fires an action with each select value change
   * 
   * @props
   * - data: array of string
   *      options to select from
   * - placeholder: string
   *      defaulted disabled option
   * - action: function
   *      callback function to call on expand while passing it the value of the selected option
   * @state
   * - 
   */
  const [value, setValue] = React.useState('');

  const onSelectChange = ({ target: { value } }) => {
    setValue(value);
    action(value);
  };

  const onCancelClick = () => {
    setValue('');
    action('');
  };

  return (
    <Grid data-testid="act-select" container spacing={2} >
      <Grid item xs={1.2}>
      <NativeSelect data-testid="select" value={value} onChange={onSelectChange} style={{fontWeight: 'bold'}}>
        <option disabled value="">
          {placeholder}
        </option>
        {data.map((elem, idx) => (
          <option key={idx} value={elem}>{elem}</option>
        ))}
      </NativeSelect>
      </Grid>
      { value != '' && <Grid item xs={10.8}> <CancelPresentationTwoToneIcon data-testid="cancel" onClick={onCancelClick} /> </Grid>}
    </Grid>
  );
};

ActionableSelect.propTypes = {
  data: PropTypes.arrayOf(PropTypes.string).isRequired,
  action: PropTypes.func.isRequired,
  placeholder: PropTypes.string.isRequired,
};

export default ActionableSelect;