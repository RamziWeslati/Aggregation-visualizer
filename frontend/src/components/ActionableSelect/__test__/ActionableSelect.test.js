import React from 'react';
import ReactDom from 'react-dom';

import { render, fireEvent } from '@testing-library/react';
import * as renderer from 'react-test-renderer';

import ActionableSelect from './../ActionableSelect';

const props = {
    data: ['option1', 'option2', 'option3'],
    action: jest.fn(),
    placeholder: 'placeholder text',
}

// cleanup is automated

// uts
it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDom.render(<ActionableSelect {...props} />, div)
})

it('renders placeholder on mount', () => {
    const { getByTestId } = render(<ActionableSelect {...props} />)
    expect(getByTestId('act-select')).toHaveTextContent('placeholder text')
})

it('does not render a cancel button on mount', () => {
    const { queryByTestId  } = render(<ActionableSelect {...props} />)
    expect(queryByTestId('cancel')).toBeNull();
})

it(`renders ${props.data.length} options on select click`, () => {
    const tree = render(<ActionableSelect {...props} />)
    const select = tree.getByTestId('select')
    fireEvent.click(select)
    expect(tree.getAllByRole('option')).toHaveLength(props.data.length + 1) // +1 for the placeholder disabled option
})


// snapshots
it('matches snapshot', () => {
    const tree = renderer.create(<ActionableSelect {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
})
