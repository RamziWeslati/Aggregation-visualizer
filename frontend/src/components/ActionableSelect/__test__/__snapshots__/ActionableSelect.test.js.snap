// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
  data-testid="act-select"
>
  <div
    className="MuiGrid-root MuiGrid-item"
  >
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline"
      data-testid="select"
      onClick={[Function]}
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      <select
        className="MuiNativeSelect-root MuiNativeSelect-select MuiInputBase-input MuiInput-input"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        value=""
      >
        <option
          disabled={true}
          value=""
        >
          placeholder text
        </option>
        <option
          value="option1"
        >
          option1
        </option>
        <option
          value="option2"
        >
          option2
        </option>
        <option
          value="option3"
        >
          option3
        </option>
      </select>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiNativeSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;
